tests/floating-point/fft/implementations/fft-pipelined.fil

```diff
@@ -0,0 +1,149 @@
+import "primitives/signed.fil";
+import "helpers.fil";
+import "primitives/state.fil";
+import "primitives/reshape.fil";
+
+/// Implements a combinational PEASE fft dataflow.
+/// Uses `N` butterflies
+comp CombPeaseFFT[NStages, N, W, E, ?M=W-E-1]<'G: L>(
+  go: interface['G],
+  in[NPoints*2]: ['G, 'G+1] W
+) -> (
+  out[NPoints*2]: ['G+L, 'G+L+1] W
+) with {
+  let NPoints = pow2(NStages);
+  exists L where L > 0;
+} where
+  // number of butterflies (N) must divide the total number of butterflies per stage (NPoints/2)
+  NPoints % (2*N) == 0,
+  N > 0,
+  NStages > 0,
+  W == E + M + 1,
+  E > 1,
+  M > 1
+{
+  assume pow2(NStages-1) == NPoints/2;
+  assume pow2(NStages-2) == NPoints/4;
+
+  twiddle := new TwiddleFactor[NPoints]<'G>();
+  bitrev := new BitRev[NStages, W]<'G>(in{0..NPoints*2});
+
+  // Creates a bundle that has NStages blocks of data,
+  // where each block has the right timings.
+  // would be nicer with 2d bundles...
+  bundle data[(NStages+1) * NPoints * 2]: for<i> ['G+i/(NPoints*2) * stage_latency, 'G + i/(NPoints*2) * stage_latency + 1] W;
+
+  data{0..NPoints*2} = bitrev.out{0..NPoints*2};
+
+  // butterfly components
+  B := new Butterflies[N, W, E, M];
+  // number of times we have to reuse the butterflies per stage
+  let reuses = NPoints / (2*N);
+  let stage_latency = B::L * reuses;
+
+  for stage in 0..NStages {
+    // which chunk of data we are populating
+    let stage_chunk = stage * NPoints * 2; 
+    // what time this chunk occurs at
+    let stage_time = stage * stage_latency;
+    for i in 0..reuses {
+      assume i*4*N < NPoints*2;
+      let start = stage_chunk + i*4*N;
+      let end = stage_chunk + (i+1)*4*N;
+      assume start / (NPoints*2) == stage;
+      assume end / (NPoints*2) == stage;
+      b := B<'G + stage_time + i * B::L>(data{start..end}, twiddle.out{i*2*N..(i+1)*2*N});
+
+      // concat into one wire for the register
+      cb := new ConcatBundle[W, 4*N]<'G + stage_time + (i + 1) * B::L>(b.out{0..4*N});
+      // register to hold values till output time
+      // TODO (edmund): This register can be reused per stage
+      reg := new Register[W*4*N]<'G + stage_time + (i + 1) * B::L, 'G + stage_time + stage_latency + 1>(cb.out);
+      // split back into bundle
+      sp := new SplitWire[W, 4*N]<'G + stage_time + stage_latency>(reg.out);
+
+      // Write to the next stage
+      data{start + NPoints*2..end + NPoints*2} = sp.out{0..4*N};
```

Comment:
Very surprisingly, commenting this line out makes the SMT solver terminate quickly. I've tried adding more assumptions like `end - start == 4*N` but that doesn't seem to help.

Reply:
Actually, maybe this is not that surprising give that `data` has a really complex liveness type and `sp` probably gets a really complicated one when `stage_time + stage_latency` is expanded.

Reply:
Okay, @UnsignedByte I've added a new assumption that seems to help. Note that only `z3` seems to terminate with these changes so cvc5 still won't work. The problem still seems to arise from the use of `/` which is maybe unsurprising. This makes a pretty strong case for #351.

Reply:
Thanks! Still not passing on `z3` for me on the ssh machine, which is weird. I'll take another look.